Feature/move sta ifaces from beerocks agent conf to platform db

adam1985d opened the pull request:
The following changes are contained in this PR:
- add generic BPL API to read prplmesh radio param from UCI
- add BPL APIs to read sta_iface from BPL (the UGW utilizes the API above)
- change sta_iface reads from beerocks_agent.conf to read using BPL
- platform_mgr is now responsible to read the sta_ifaces and update the config struct
- configurations that depend on sta_iface have moved to a later stage (after BPL is initialized and sta_ifaces are read from BPL)
- sta_iface was removed from beerocks_agent.conf and the sta_iface was removed from the config_file::read_slave_config_file().

adam1985d commented:
![](https://github.trello.services/images/mini-trello-icon.png) [move sta_iface from beerocks_agent.conf to bpl](https://trello.com/c/UEflUZjg/49-move-staiface-from-beerocksagentconf-to-bpl)

tomereli commented on framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp:
```diff
@@ -49,6 +49,22 @@ int cfg_get_prplmesh_param(const std::string &param, char *buf, size_t buf_len)
+        buf_len = MAX_UCI_BUF_LEN;
+    }
+
+    if (snprintf_s(path, MAX_UCI_BUF_LEN, "prplmesh.radio%d.%s", radio_id, radio_param.c_str()) <=
```
I personally don't like adding '{' for one liners, @morantr thinks otherwise, but whatever you choose please be consistent.
(nitpick)

adam1985d replied:
added `{` and `}` as I agree with @morantr on this

tomereli commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -456,6 +456,17 @@ int cfg_notify_iface_status(const BPL_INTERFACE_STATUS_NOTIFICATION *status_noti
+ * @return 0 Success.
+ * @return -1 Error.
+ */
+int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAME_LEN]);
```
I think this should have been using explicit size for strings:
```
int cfg_get_sta_iface(const char *iface, size_t iface_length, char *sta_iface, size_t sta_iface_length);
```

Since this is not the case in the whole BPL UCI implementation - (nitpick)

adam1985d replied:
skipped for now

arnout replied:
> I think this should have been using explicit size for strings:

Actually it should have used std::string.

> Since this is not the case in the whole BPL UCI implementation - (nitpick)

This point remains of course.

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -220,5 +220,16 @@ int cfg_notify_iface_status(const BPL_INTERFACE_STATUS_NOTIFICATION *status_noti
+int cfg_get_sta_iface(const char iface[BPL_IFNAME_LEN], char sta_iface[BPL_IFNAME_LEN])
+{
+    if (iface == NULL || sta_iface == NULL) {
+        return RETURN_ERR;
```
Why not add log(ERR) here?

adam1985d replied:
added

tomereli commented on framework/platform/bpl/linux/bpl_cfg.cpp:
```diff
@@ -220,5 +220,16 @@ int cfg_notify_iface_status(const BPL_INTERFACE_STATUS_NOTIFICATION *status_noti
+}
+
 } // namespace bpl
 } // namespace beerocks
```
Can you add empty line here before end of file?

adam1985d replied:
added

tomereli commented on agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:
```diff
@@ -1664,5 +1664,20 @@ bool main_thread::init_dhcp_monitor()
     return true;
 }
 
+std::string main_thread::get_sta_iface(int slave_num)
```
If you are not accessing other class parameters here, better make this static instead of class private.

adam1985d replied:
but I am accessing the `config.hostap_iface[]` which is a class parameter.

I changed the API to get hostap_iface and the API is now static.

I changed the corresponding use in the following commit titled "platform_manager - add sta_ifaces to iface_status_map in init"

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -149,8 +149,20 @@ static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_
-    son_slave_conf.backhaul_wireless_iface = beerocks_slave_conf.sta_iface[slave_num];
+    son_slave_conf.radio_identifier = beerocks_slave_conf.radio_identifier[slave_num];
+    // read the sta_iface from bpl and verify it is available
+    char sta_iface_str[BPL_IFNAME_LEN];
```
Please move this to a function to keep the main small. The function should be generic if possible since some of the other parameters here will also move to BPL as well.

adam1985d replied:
ok - moved to a separate static function

tomereli commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -311,6 +303,15 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
+        // The platform manager updates the beerocks_slave_conf.sta_iface in the init stage
+        std::set<std::string> slave_sta_ifaces;
+        for (int slave_num = 0; slave_num < beerocks::IRE_MAX_SLAVES; slave_num++) {
+
```
Remove extra line break

adam1985d replied:
done

tomereli reviewed (approved):
commit 1d400004d37117534df42b0a022915e574bf1a4c - commit message has some typos and its duplicated. Please fix that.

Other than that, excellent work!!! Great split to commits. Commit messages can be slightly improved.
You will need to also create corresponding PRs in UGW and prplWrt feed.
Please update the PR description with links to the UGW PR which adds the STA iface to /etc/config/prplmesh in the feed (its OK that only Intel can access the link).

Please also do a similar change here - https://git.prpl.dev/intel/iwlwav/blob/iwlwav-19.07/prplmesh/files/etc/config/prplmesh - this is the prplWrt used for RAX40. This can be done in a separate, followup PR.

adam1985d reviewed (commented):
> commit [1d40000](https://github.com/prplfoundation/prplMesh/commit/1d400004d37117534df42b0a022915e574bf1a4c) - commit message has some typos and its duplicated. Please fix that.
**done**

> 
> Other than that, excellent work!!! Great split to commits. Commit messages can be slightly improved.
> You will need to also create corresponding PRs in UGW and prplWrt feed.
> Please update the PR description with links to the UGW PR which adds the STA iface to /etc/config/prplmesh in the feed (its OK that only Intel can access the link).

**will be updated - after the review is complete and I will have the correct hash ID**
> 
> Please also do a similar change here - https://git.prpl.dev/intel/iwlwav/blob/iwlwav-19.07/prplmesh/files/etc/config/prplmesh - this is the prplWrt used for RAX40. This can be done in a separate, follow-up PR.
**there is no need to change the /etc/config/prplmesh file since the default configuration is GW only and sta_iface disabled from prplmesh perspective**

adam1985d committed:
bpl:cfg_get_prplmesh_radio_param - add new API

The existing implementation of BPL supports reading
only parameters that are in the config section
in the UCI.

The new API provides an interface to get parameters
configured under the radio section in the UCI.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
bpl:cfg_get_sta_iface - add new API to get sta_iface

The new API receives an interface name as incoming param
and a char array as an outgoing parameter for the result.
If an error occurred or no sta_iface is set, the API will
return -1, otherwise 0.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
platform_manager:get_sta_iface - get sta iface from slave num

The new private method provided utilizes the new introduced BPL
API to get sta_iface from the BPL instead of the beerocks_agent.conf

If the operation failed or the interface doesn't exist, the method will
return an empty string.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
platform_manager - add sta_ifaces to iface_status_map in init

Before the change, the sta_ifaces were aded to the bpl_iface_status_map
at the main_thread() function (CTOR).

After the move of sta_iface param from beerocks_agent.conf to BPL, the
sta_iface information is available only after the BPL is initialized.
The initialization of BPL is done in init() and thus the insertion of
the available sta_ifaces is also done in init().

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
slave_main: set backhaul_wireless_iface from sta_iface in BPL

The backhaul_wireless_iface in the son_slave_conf was set from
the beerocks_slave_conf.sta_iface - a value which is read from
the beerocks_agent.conf file.

After this change, the backhaul_wireless_iface is set from the
sta_iface value read from BPL.
As part of this change, the point of verification of the Linux
interface was moved as well.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
slave_main: set slave_sta_ifaces after platform_mgr init

The sta_iface param for each agent is updated as part of the
platform_mgr init(). This means that the dependant slave_sta_ifaces
set can only be created after the platform_mgr init() is complete.

Signed-off-by: Adam Dov <adamx.dov@intel.com>

adam1985d committed:
beerocks_agent.conf: remove sta_iface param

The sta_iface param is moved to the BPL. As part of this change,
the sta_iface is removed from the beerocks_agent.conf file and from the
beerocks_config_file:read_slave_config_file().

Signed-off-by: Adam Dov <adamx.dov@intel.com>

arnout commented on framework/platform/bpl/uci/cfg/bpl_cfg_helper.cpp:
```diff
@@ -49,6 +49,23 @@ int cfg_get_prplmesh_param(const std::string &param, char *buf, size_t buf_len)
+        buf_len = MAX_UCI_BUF_LEN;
+    }
+
+    if (snprintf_s(path, MAX_UCI_BUF_LEN, "prplmesh.radio%d.%s", radio_id, radio_param.c_str()) <=
```
Not for this PR, but: why is this file using snprintf_s instead of a sstream?

Even more unrelated: I think it would be convenient if we would import the [fmt library](https://github.com/fmtlib/fmt) and use the C++20 formatter.

tomereli replied:
This will be a great addition! I always wondered why c++ fails to have something like this as part of the standard library.

arnout commented on framework/platform/bpl/include/bpl/bpl_cfg.h:
```diff
@@ -456,6 +456,17 @@ int cfg_notify_iface_status(const BPL_INTERFACE_STATUS_NOTIFICATION *status_noti
+ * @param [out] sta_iface name of STA interface (up to 32 bytes in length).
+ *
+ * @return 0 Success.
+ * @return -1 Error.
```
It could be clarified here that it also returns -1 if there is no STA interface configured on that radio. (nitpick)

arnout commented on agent/src/beerocks/slave/platform_manager/platform_manager_thread.cpp:
```diff
@@ -281,6 +281,21 @@ std::string extern_query_db(std::string parameter)
+{
+    char sta_iface_str[BPL_IFNAME_LEN];
+    if (bpl::cfg_get_sta_iface(hostap_iface.c_str(), sta_iface_str) < 0) {
+        LOG(ERROR) << "failed to read sta_iface for slave ";
```
I don't think this is really an error - it's perfectly fine for a radio not to have a STA configured, and the only way to know whether or not it has one configured is by calling this function. (nitpick)

arnout commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -125,6 +125,25 @@ static bool parse_arguments(int argc, char *argv[])
+        return std::string();
+    } else {
+        sta_iface = std::string(sta_iface_str);
+        if (!beerocks::net::network_utils::linux_iface_exists(sta_iface)) {
+            LOG(DEBUG) << "sta iface " << sta_iface << " does not exist, clearing it from config";
+            sta_iface.clear();
+        }
```
I think this should move into bpl::cfg_get_sta_iface. It is possible that the platform completely removes the interface when the STA is disabled, so linux_iface_exists will fail if the STA link is not up. So for those platforms, the bpl should take care of validating the correctness in a different way. Also, for example, for dummy bpl we don't really care if the interface exists.

Obviously this can be done in a future PR. I don't even think it's necessary to make an issue for it - we'll handle it if the problem presents itself (i.e. when doing test_flows with wireless backhaul).

arnout commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -149,8 +168,9 @@ static void fill_son_slave_config(beerocks::config_file::sConfigSlave &beerocks_
-    son_slave_conf.backhaul_wireless_iface = beerocks_slave_conf.sta_iface[slave_num];
+    son_slave_conf.radio_identifier = beerocks_slave_conf.radio_identifier[slave_num];
+    son_slave_conf.backhaul_wireless_iface =
+        get_sta_iface_from_hostap_iface(son_slave_conf.hostap_iface);
```
Hm, I'm not 100% sure if this is OK. bpl supposedly should only be used from the platform_manager thread (which BTW is going to become a separate process - cfr. #575). It is still OK-ish to do it from the main process, but I'm a bit concerned about adding calls to the son_slave - currently there aren't any.

So I think this should instead be part of the wlan_settings in platform_settings and sent by the platform manager in fill_platform_settings.

Clearly this is something that we can handle later.

arnout reviewed (approved):
Nice PR. I have comments but nothing blocking merge.

I especially like that the commits are ordered so that nothing ever breaks, and that you added doxygen documentation.

arnout commented:
I fixed my nitpicks and rebased on master.
